components/calendar/src/islamic.rs

```diff
@@ -0,0 +1,529 @@
+
+    fn offset_date(&self, date: &mut Self::DateInner, offset: DateDuration<Self>) {
+        date.0.offset_date(offset)
+    }
+
+    fn until(
+        &self,
+        date1: &Self::DateInner,
+        date2: &Self::DateInner,
+        _calendar2: &Self,
+        _largest_unit: DateDurationUnit,
+        _smallest_unit: DateDurationUnit,
+    ) -> DateDuration<Self> {
+        date1.0.until(date2.0, _largest_unit, _smallest_unit)
+    }
+
+    fn debug_name(&self) -> &'static str {
+        "IslamicObservational"
+    }
+
+    fn year(&self, date: &Self::DateInner) -> types::FormattableYear {
+        Self::year_as_islamic(date.0.year)
+    }
+
+    fn month(&self, date: &Self::DateInner) -> types::FormattableMonth {
+        date.0.solar_month()
+    }
+
+    fn day_of_month(&self, date: &Self::DateInner) -> types::DayOfMonth {
+        date.0.day_of_month()
+    }
+
+    fn day_of_year_info(&self, date: &Self::DateInner) -> types::DayOfYearInfo {
+        let prev_year = date.0.year.saturating_sub(1);
+        let next_year = date.0.year.saturating_add(1);
+        types::DayOfYearInfo {
+            day_of_year: date.0.day_of_year(),
+            days_in_year: date.0.days_in_year(),
+            prev_year: Self::year_as_islamic(prev_year),
+            days_in_prev_year: Self::days_in_provided_year(prev_year),
+            next_year: Self::year_as_islamic(next_year),
+        }
+    }
+    // TODO: ADD TO ANYCALENDAR
+    // fn any_calendar_kind(&self) -> Option<AnyCalendarKind> {
+    //     Some(AnyCalendarKind::IslamicObservational)
+    // }
+}
+
+impl IslamicObservational {
+    /// Constructs a new Islamic Observational Calendar
+    pub fn new() -> Self {
+        Self
+    }
+
+    // "Fixed" is a day count representation of calendars staring from Jan 1st of year 1 of the Georgian Calendar.
+    // The fixed date algorithms are from
+    // Dershowitz, Nachum, and Edward M. Reingold. _Calendrical calculations_. Cambridge University Press, 2008.
+    //
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6904
+    fn fixed_from_islamic(i_date: IslamicDateInner) -> RataDie {
+        let year = i64::from(i_date.0.year);
+        let month = i64::from(i_date.0.month);
+        let day = i64::from(i_date.0.day);
+
+        let midmonth = FIXED_ISLAMIC_EPOCH_FRIDAY.to_f64_date()
+            + (((year - 1) as f64) * 12.0 + month as f64 - 0.5) * MEAN_SYNODIC_MONTH;
+
+        Astronomical::phasis_on_or_before(RataDie::new(midmonth as i64), CAIRO) + day - 1
+    }
+
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6920
+    #[allow(clippy::unwrap_used)]
+    fn islamic_from_fixed(date: RataDie) -> Date<IslamicObservational> {
+        let crescent = Astronomical::phasis_on_or_before(date, CAIRO);
+        let elapsed_months =
+            (libm::round((crescent - FIXED_ISLAMIC_EPOCH_FRIDAY) as f64 / MEAN_SYNODIC_MONTH))
+                as i32;
```

Comment:
Elapsed months is an i32 because it is used in the calculation for year, which needs to be an i32. I also div_rem_euclid by 12 because the lisp code uses the quotient and mod functions which correspond to euclidean division.